### PR TITLE
Unity 3.4.2 Adapter Update

### PR DESCRIPTION
**Pull request by jindoujiao**
Update Admob Unity Adapter.

### CODE REVIEW

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 1]
```diff
@@ -122,6 +125,12 @@ public void onUnityAdsPlacementStateChanged(String placementId,
             // This callback is not forwarded to the adapter by the UnitySingleton and the
             // adapter should use the onUnityAdsReady and onUnityAdsError callbacks to forward
             // Unity Ads SDK state to Google Mobile Ads SDK.
+
+            //todo: no fill
```
Comment (thatguyhampton) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: Can probably remove this comment

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 2]
```diff
@@ -273,38 +324,13 @@ public void requestBannerAd(Context context,
...
-            bannerListener.onAdFailedToLoad(UnityAdapter.this, AdRequest.ERROR_CODE_NO_FILL);
-        } else {
-            bannerListener.onAdFailedToLoad(UnityAdapter.this, AdRequest.ERROR_CODE_INTERNAL_ERROR);
-        }
-    }
-
-    @Override
-    public void onBannerLeftApplication(BannerView bannerView) {
-        bannerListener.onAdLeftApplication(UnityAdapter.this);
+        return null;
```
Comment (thatguyhampton) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: Should this be returning `mBannerView`?

**Review by thatguyhampton — CHANGES_REQUESTED** [review]

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 3]
```diff
@@ -119,31 +121,69 @@ public void onUnityAdsClick(String s) {
...
 
         @Override
         public void onUnityAdsFinish(String placementId, UnityAds.FinishState finishState) {
             // Unity Ads ad closed.
-            if (mMediationInterstitialListener != null) {
-                mMediationInterstitialListener.onAdClosed(UnityAdapter.this);
+            if (placementId.equals(getPlacementId())) {
+                if (mMediationInterstitialListener != null) {
+                    if (finishState == UnityAds.FinishState.ERROR) {
+                        mMediationInterstitialListener.onAdFailedToLoad(UnityAdapter.this, AdRequest.ERROR_CODE_INTERNAL_ERROR);
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: according to docs, `mMediationInterstitialListener.onAdFailedToLoad` should be used only when we failed to load ad.

Comment (serge-14) [reply ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: It looks like we should just ignore error here and do nothing. Also we should check with AdMob team.

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 4]
```diff
@@ -193,57 +234,71 @@ public void requestInterstitialAd(Context context,
...
 
-        UnitySingleton.getInstance().initializeUnityAds(activity, gameId);
-        UnitySingleton.getInstance().loadAd(mUnityAdapterDelegate);
-
+        UnityAds.addListener(mUnityAdapterDelegate);
+        UnitySingleton.getInstance().initializeUnityAds(activity, gameId,
+                new UnitySingleton.Listener() {
+                    @Override
+                    public void onInitializeSuccess() {
+                        UnityAds.load(mPlacementId);
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: why can't we call load right after init?

```
UnitySingleton.getInstance().initializeUnityAds(...);
UnityAds.load(mPlacementId);
```

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 5]
```diff
@@ -193,57 +234,71 @@ public void requestInterstitialAd(Context context,
...
-        mMediationInterstitialListener.onAdOpened(UnityAdapter.this);
-
         if (mActivityWeakReference != null && mActivityWeakReference.get() != null) {
-            // Request UnitySingleton to show interstitial ads.
-            UnitySingleton.getInstance().showAd(mUnityAdapterDelegate, mActivityWeakReference.get());
+           UnityAds.show(mActivityWeakReference.get(), mPlacementId);
         } else {
             Log.w(TAG, "Failed to show Unity Ads Interstitial.");
-            mMediationInterstitialListener.onAdClosed(UnityAdapter.this);
+            mMediationInterstitialListener.onAdFailedToLoad(UnityAdapter.this, AdRequest.ERROR_CODE_INVALID_REQUEST);
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: Same comment as above for `mMediationInterstitialListener.onAdFailedToLoad`

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityMediationAdapter.java [hunk 6]
```diff
@@ -57,6 +57,12 @@
      */
     private String mPlacementId;
 
+    /**
+     * Used by Unity Ads to track failures in the mediation lifecycle
+     */
+    private int impressionOrdinal;
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityMediationAdapter.java]: looks like we are not using those.

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnitySingleton.java [hunk 7]
```diff
@@ -15,35 +15,19 @@
...
  * The {@link UnitySingleton} class is used to load {@link UnityAds}, handle multiple
  * {@link UnityAdapter} instances and mediate their callbacks.
  */
 public final class UnitySingleton {
-
-    /**
-     * A list of adapter listeners with their respective placement IDs to prevent duplicate
-     * requests.
-     */
-    private HashMap<String, WeakReference<UnityAdapterDelegate>> mPlacementsInUse;
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnitySingleton.java]: why do we remove it?

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 8]
```diff
@@ -190,60 +242,95 @@ public void requestInterstitialAd(Context context,
...
+                new UnitySingleton.Listener() {
+                    @Override
+                    public void onInitializeSuccess() {
+                        MetaData metadata = new MetaData(activity);
+                        metadata.setCategory("mediation_adapter");
+                        metadata.set(uuid, "load-interstitial");
+                        metadata.set(uuid, mPlacementId);
+                        metadata.commit();
+
+                        UnitySingleton.getInstance().mPlacementsInUse.add(mPlacementId);
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java]: should we set it before calling `initializeUnityAds`? also if it is in we should report no fill immideatly to be aligned with previous implementation.

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java [hunk 9]
```diff
@@ -190,60 +242,95 @@ public void requestInterstitialAd(Context context,
...
+            mMediationInterstitialListener.onAdOpened(UnityAdapter.this);
             mMediationInterstitialListener.onAdClosed(UnityAdapter.this);
         }
     }
+    //endregion
 
+    //region MediationAdapter implementation.
     @Override
     public void onDestroy() {
+        MetaData metadata = new MetaData((mActivityWeakReference.get()));
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java:294]: should we remove placement id from  `UnitySingleton.getInstance().mPlacementsInUse`?

File: ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityMediationAdapter.java [hunk 10]
```diff
@@ -285,6 +337,10 @@ private static boolean isValidIds(String gameId, String placementId) {
 
             return false;
         }
+        if (UnitySingleton.getInstance().mPlacementsInUse.contains(placementId)) {
```
Comment (serge-14) [inline ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityMediationAdapter.java]: this is will report invalid request error... old implementation had no fill.

**Pull request merged**